File: sdk/go/zip321.go
package zns

import (
	"encoding/base64"
	"fmt"
	"net/url"
	"strings"
)

// Zip321Parts holds the parsed components of a ZIP-321 zcash: URI.
type Zip321Parts struct {
	Address     string
	Amount      string
	MemoRaw     string
	MemoDecoded string
}

// ToBase64URL encodes a string to base64url (RFC 4648 section 5) without padding.
func ToBase64URL(text string) string {
	return base64.RawURLEncoding.EncodeToString([]byte(text))
}

// DecodeBase64URL decodes a base64url-encoded string (with or without padding).
// Returns an empty string if decoding fails.
func DecodeBase64URL(value string) string {
	// Try without padding first (RawURLEncoding)
	decoded, err := base64.RawURLEncoding.DecodeString(value)
	if err != nil {
		// Try with padding (URLEncoding)
		decoded, err = base64.URLEncoding.DecodeString(value)
		if err != nil {
			return ""
		}
	}
	return string(decoded)
}

// BuildZcashURI constructs a ZIP-321 zcash: payment URI.
// The memo should be a plain text string; it will be base64url-encoded automatically.
func BuildZcashURI(address string, amount string, memo string) string {
	encodedMemo := ToBase64URL(memo)
	return fmt.Sprintf("zcash:%s?amount=%s&memo=%s",
		address,
		url.QueryEscape(amount),
		url.QueryEscape(encodedMemo),
	)
}

// ParseZip321URI parses a ZIP-321 zcash: URI into its component parts.
// Returns an empty Zip321Parts if the URI is not a valid zcash: URI.
func ParseZip321URI(uri string) Zip321Parts {
	// Must start with "zcash:"
	if !strings.HasPrefix(uri, "zcash:") {
		return Zip321Parts{}
	}

	// Split off the "zcash:" prefix
	rest := strings.TrimPrefix(uri, "zcash:")

	// Split address from query params
	var address, queryStr string
	if idx := strings.Index(rest, "?"); idx >= 0 {
		address = rest[:idx]
		queryStr = rest[idx+1:]
	} else {
		address = rest
		return Zip321Parts{Address: address}
	}

	params, err := url.ParseQuery(queryStr)
	if err != nil {
		return Zip321Parts{Address: address}
	}

	amount := params.Get("amount")
	memoRaw := params.Get("memo")
	memoDecoded := DecodeBase64URL(memoRaw)

	return Zip321Parts{
		Address:     address,
		Amount:      amount,
		MemoRaw:     memoRaw,
		MemoDecoded: memoDecoded,
	}
}

File: sdk/swift/Sources/ZcashNameSDK/Zip321.swift
import Foundation

// MARK: - Zip321Parts

/// Parsed components of a ZIP-321 payment URI.
public struct Zip321Parts: Sendable {
    public let address: String
    public let amount: String
    public let memoRaw: String
    public let memoDecoded: String

    public init(address: String, amount: String, memoRaw: String, memoDecoded: String) {
        self.address = address
        self.amount = amount
        self.memoRaw = memoRaw
        self.memoDecoded = memoDecoded
    }
}

// MARK: - Base64URL encoding/decoding

/// Encode a UTF-8 string to Base64URL (no padding).
public func toBase64URL(_ text: String) -> String {
    guard let data = text.data(using: .utf8) else { return "" }
    return data.base64EncodedString()
        .replacingOccurrences(of: "+", with: "-")
        .replacingOccurrences(of: "/", with: "_")
        .replacingOccurrences(of: "=", with: "")
}

/// Decode a Base64URL string back to a UTF-8 string.
public func decodeBase64URL(_ value: String) -> String {
    var normalized = value
        .replacingOccurrences(of: "-", with: "+")
        .replacingOccurrences(of: "_", with: "/")

    // Add padding if needed.
    let remainder = normalized.count % 4
    if remainder != 0 {
        normalized += String(repeating: "=", count: 4 - remainder)
    }

    guard let data = Data(base64Encoded: normalized) else { return "" }
    return String(data: data, encoding: .utf8) ?? ""
}

// MARK: - URI building / parsing

/// Build a `zcash:` payment URI (ZIP-321 format).
///
/// - Parameters:
///   - address: The Zcash address.
///   - amount: Optional amount string (e.g. "0.001").
///   - memo: Optional plain-text memo (will be Base64URL-encoded).
/// - Returns: A `zcash:` URI string.
public func buildZcashURI(address: String, amount: String? = nil, memo: String? = nil) -> String {
    guard !address.isEmpty else { return "" }
    let base = "zcash:\(address)"
    var params: [String] = []
    if let amount, let amountValue = Double(amount), amountValue > 0 {
        params.append("amount=\(amount)")
    }
    if let memo, !memo.isEmpty {
        params.append("memo=\(toBase64URL(memo))")
    }
    return params.isEmpty ? base : "\(base)?\(params.joined(separator: "&"))"
}

/// Parse a ZIP-321 `zcash:` URI into its component parts.
///
/// - Parameter uri: The URI string to parse.
/// - Returns: A `Zip321Parts` struct with the extracted components.
public func parseZip321URI(_ uri: String) -> Zip321Parts {
    let raw = uri.trimmingCharacters(in: .whitespaces)
    let withoutScheme: String
    if raw.lowercased().hasPrefix("zcash:") {
        withoutScheme = String(raw.dropFirst("zcash:".count))
    } else {
        withoutScheme = raw
    }

    let parts = withoutScheme.split(separator: "?", maxSplits: 1)
    let address = String(parts.first ?? "").trimmingCharacters(in: .whitespaces)
    let queryPart = parts.count > 1 ? String(parts[1]) : ""

    let queryItems = parseQueryString(queryPart)
    let amount = (queryItems["amount"] ?? "").trimmingCharacters(in: .whitespaces)
    let memoRaw = (queryItems["memo"] ?? "").trimmingCharacters(in: .whitespaces)
    let memoDecoded = memoRaw.isEmpty ? "" : decodeBase64URL(memoRaw)

    return Zip321Parts(address: address, amount: amount, memoRaw: memoRaw, memoDecoded: memoDecoded)
}

// MARK: - Private helpers

/// Simple query string parser (avoids URLComponents dependency quirks with zcash: scheme).
private func parseQueryString(_ query: String) -> [String: String] {
    guard !query.isEmpty else { return [:] }
    var result: [String: String] = [:]
    for pair in query.split(separator: "&") {
        let kv = pair.split(separator: "=", maxSplits: 1)
        if kv.count == 2 {
            result[String(kv[0])] = String(kv[1])
        } else if kv.count == 1 {
            result[String(kv[0])] = ""
        }
    }
    return result
}
